File: internal/generators/scaffold_doctor.go
package generators

import (
	"bytes"
	"fmt"
	"os"
	"path/filepath"
	"sort"

	"github.com/kgsaran/trackfw/internal/integrations"
	"github.com/kgsaran/trackfw/internal/version"
)

// pythonValidateScriptForm is the byte-exact content Python's `trackfw init` and
// `trackfw update` (validate-script target) write to scripts/trackfw-validate.sh.
// It is accepted by the set-membership check in checkValidateScriptArtifact so that
// a project initialized by the Python runtime does not produce a false-positive
// scaffold-divergent finding in the Go doctor.
//
// Why a named constant here (not in scaffold.go): this string is only consumed by
// the doctor's membership check, not by any generator; co-locating it with the
// check avoids confusion about which form Go's generator emits.
//
// Scope of the exception: ONLY scripts/trackfw-validate.sh uses set-membership.
// All other scaffold artifacts are compared against the single template the local
// runtime would generate (exact bytes). See docs/cli-parity.md,
// "validate.sh — pertencimento a conjunto (set-membership, escopado)".
const pythonValidateScriptForm = "#!/usr/bin/env bash\nset -euo pipefail\ntrackfw validate\n"

// RunScaffoldDoctor compares scaffold artifacts on disk against the templates the
// currently installed binary would generate (given the project's own trackfw.yaml),
// and returns findings for any artifact that is divergent or missing.
//
// Design decisions (ADR-2026-08-27, REQ-2026-08-27):
//
//   - Property by path, not by manifest (AC3): scaffold artifacts are identified by
//     well-known namespace paths (.claude/commands/trackfw/, scripts/trackfw-*.sh,
//     .github/workflows/trackfw-gate.yml). No manifest entry is written or read.
//
//   - Sibling classifier (AC15): scaffold artifacts are never in the manifest, so
//     routing them through ClassifyDoctor would produce a meaningless Claim and a wrong
//     remedy. The finding kinds DoctorScaffoldDivergent / DoctorScaffoldMissing carry a
//     zero Claim and a trackfw-update remedy, distinct from the catalog-based kinds.
//
//   - Config-rendered templates (AC12): scripts/trackfw-validate.sh content varies with
//     cfg.Backend/cfg.Frontend — the template is rendered from the project's own
//     trackfw.yaml, not from a hardcoded default. Any project with backend: go would be a
//     false positive otherwise.
//
//   - validate.sh set-membership (architect decision 2026-08-27): scripts/trackfw-validate.sh
//     is accepted when it matches ANY known runtime's template (Go/Node form rendered from
//     the project's cfg, OR Python's fixed form). This is the only artifact with this
//     exception — the byte-divergence between runtimes is pre-existing, intentional, and
//     documented. A file that matches NO known form is still accused. See
//     checkValidateScriptArtifact and docs/cli-parity.md.
//
//   - Eligibility for missing (AC14): slash commands are only checked when
//     .claude/commands/trackfw/ already exists. Its absence signals a project initialized
//     via `trackfw discover --init` (which legitimately omits slash commands) — reporting
//     9 missing files would be false positives for that initializer.
//
//   - Conditional artifacts (AC13): CI workflow is only checked when cfg.CI declares it.
//     Absence of an unconfigured artifact is never a finding.
//
//   - Neutral blame message (AC16): no scaffold artifact carries a version stamp, so the
//     binary cannot determine whether it or the project is stale. The remedy and message
//     name the installed binary version but instruct the user to verify the direction.
//
//   - Guards are included (additive coverage): credential-guard and git-branch-guard are
//     covered here in addition to the two `validate` rules that already check them.
//     Neither service is exclusively owned by the other surface — this is complementary.
//
//   - Hook files (husky/lefthook) are excluded per the declared residual in
//     docs/seguranca/2026-08-27-modelo-de-ameaca-da-cobertura-de-scaffold.md §Residual-3.
//
// AC15 note: the blocker cited in the roadmap ("ClassifyDoctor has no case for
// !Registered && StateModified") is false for all three CLIs. Go received the fix in
// ML-2C (see doctor.go line 118-124); Node.js and Python have equivalent branches. This
// function provides scaffold coverage via a separate path — not by adding a case to
// ClassifyDoctor, which would produce wrong remedies and a meaningless Claim.
func RunScaffoldDoctor(projectRoot string) ([]integrations.DoctorFinding, error) {
	// Eligibility: trackfw.yaml must exist. Without it there is no evidence this is a
	// trackfw project, so return empty findings rather than flooding a non-trackfw repo.
	if _, err := os.Stat(filepath.Join(projectRoot, "trackfw.yaml")); err != nil {
		return []integrations.DoctorFinding{}, nil
	}

	// Load config from the project's trackfw.yaml to render cfg-dependent templates
	// (AC12). config.Load() reads relative to cwd — chdir like Update() does.
	orig, err := os.Getwd()
	if err != nil {
		return nil, fmt.Errorf("scaffold doctor: getwd: %w", err)
	}
	if err := os.Chdir(projectRoot); err != nil {
		return nil, fmt.Errorf("scaffold doctor: chdir %s: %w", projectRoot, err)
	}
	defer os.Chdir(orig) //nolint:errcheck

	cfg := loadUpdateConfig()

	findings := []integrations.DoctorFinding{}

	// --- Scripts (always in scope when trackfw.yaml is present) ---
	//
	// All five scripts are written by trackfw init and trackfw update unconditionally.
	// trackfw discover --init also writes them (InstallGates), so their presence is
	// expected in any trackfw-managed project.
	//
	// scripts/trackfw-validate.sh is handled separately via checkValidateScriptArtifact
	// (set-membership against all known runtime forms). The four remaining scripts use
	// single-template equality via checkScaffoldArtifact.
	if f := checkValidateScriptArtifact(filepath.Join(projectRoot, "scripts/trackfw-validate.sh"), "scripts/trackfw-validate.sh", cfg); f != nil {
		findings = append(findings, *f)
	}
	staticScripts := []struct {
		relPath string
		content []byte
	}{
		{"scripts/trackfw-attention-signal.sh", []byte(attentionSignalScript)},
		{"scripts/trackfw-attention-cleanup.sh", []byte(attentionCleanupScript)},
		{"scripts/trackfw-credential-guard.sh", []byte(credentialGuardScript)},
		{"scripts/trackfw-git-branch-guard.sh", []byte(gitBranchGuardScript)},
	}
	for _, s := range staticScripts {
		path := filepath.Join(projectRoot, s.relPath)
		f := checkScaffoldArtifact(path, s.relPath, s.content, true)
		if f != nil {
			findings = append(findings, *f)
		}
	}

	// --- Slash commands (AC14: only when the directory already exists) ---
	//
	// The directory's presence is the eligibility signal: a project initialized via
	// `trackfw discover --init` (which does NOT write slash commands) will not have this
	// directory, so we report no missing commands. A project initialized via `trackfw
	// init` or `trackfw update` will have it, and any absent file inside is a finding.
	claudeDir := filepath.Join(projectRoot, ClaudeCommandsDirPath)
	if _, err := os.Stat(claudeDir); err == nil {
		for filename, content := range claudeCommandsContent() {
			relPath := ClaudeCommandsDirPath + "/" + filename
			path := filepath.Join(projectRoot, relPath)
			f := checkScaffoldArtifact(path, relPath, []byte(content), true)
			if f != nil {
				findings = append(findings, *f)
			}
		}
	}

	// --- CI workflow (AC13: conditional on ci: in trackfw.yaml) ---
	switch cfg.CI {
	case "github-actions":
		relPath := GitHubActionsWorkflowPath
		path := filepath.Join(projectRoot, relPath)
		f := checkScaffoldArtifact(path, relPath, []byte(buildGitHubActionsWorkflowContent(cfg)), true)
		if f != nil {
			findings = append(findings, *f)
		}
	case "gitlab-ci":
		relPath := GitLabCIWorkflowPath
		path := filepath.Join(projectRoot, relPath)
		f := checkScaffoldArtifact(path, relPath, []byte(buildGitLabCIWorkflowContent(cfg)), true)
		if f != nil {
			findings = append(findings, *f)
		}
	}

	// Deterministic output (AC7): sort by destination so the three CLIs are byte-identical
	// when diffed (scripts/check-doctor-parity.sh extension required — ML-2A).
	sort.Slice(findings, func(i, j int) bool {
		return findings[i].Destination < findings[j].Destination
	})

	return findings, nil
}

// checkValidateScriptArtifact checks scripts/trackfw-validate.sh using set-membership:
// the file is accepted if its content matches ANY known runtime's template —
// either the Go/Node form (cfg-rendered from the project's trackfw.yaml) or
// Python's fixed form (pythonValidateScriptForm). All other scaffold artifacts use
// single-template equality via checkScaffoldArtifact.
//
// Why set-membership here: the three runtimes intentionally emit different bytes for
// this file (Go/Node: shebang #!/usr/bin/env sh + cfg-dependent build steps; Python:
// simple #!/usr/bin/env bash form). The divergence is pre-existing and documented.
// A file that matches NONE of the known forms is still accused (AC3 preserved).
func checkValidateScriptArtifact(path, relPath string, cfg Config) *integrations.DoctorFinding {
	actual, err := os.ReadFile(path)
	if os.IsNotExist(err) {
		f := integrations.DoctorFinding{
			FindingKind: integrations.DoctorScaffoldMissing,
			Destination: relPath,
			Remedy:      scaffoldRemedy("restore", relPath),
		}
		return &f
	}
	if err != nil {
		f := integrations.DoctorFinding{
			FindingKind: integrations.DoctorScaffoldDivergent,
			Destination: relPath,
			Remedy:      scaffoldRemedy("resync", relPath),
		}
		return &f
	}
	goNodeForm := []byte(buildValidateScript(cfg))
	pythonForm := []byte(pythonValidateScriptForm)
	if bytes.Equal(actual, goNodeForm) || bytes.Equal(actual, pythonForm) {
		return nil
	}
	f := integrations.DoctorFinding{
		FindingKind: integrations.DoctorScaffoldDivergent,
		Destination: relPath,
		Remedy:      scaffoldRemedy("resync", relPath),
	}
	return &f
}

// checkScaffoldArtifact compares the on-disk content at path against expected.
// Returns a finding if the file is divergent or (when reportMissing=true) absent.
// relPath is used as Destination in the finding (relative to project root, human-readable).
func checkScaffoldArtifact(path, relPath string, expected []byte, reportMissing bool) *integrations.DoctorFinding {
	actual, err := os.ReadFile(path)
	if os.IsNotExist(err) {
		if !reportMissing {
			return nil
		}
		f := integrations.DoctorFinding{
			FindingKind: integrations.DoctorScaffoldMissing,
			Destination: relPath,
			Remedy:      scaffoldRemedy("restore", relPath),
		}
		return &f
	}
	if err != nil {
		// Unreadable artifact: treat as divergent so the user is informed.
		f := integrations.DoctorFinding{
			FindingKind: integrations.DoctorScaffoldDivergent,
			Destination: relPath,
			Remedy:      scaffoldRemedy("resync", relPath),
		}
		return &f
	}
	if bytes.Equal(actual, expected) {
		return nil
	}
	f := integrations.DoctorFinding{
		FindingKind: integrations.DoctorScaffoldDivergent,
		Destination: relPath,
		Remedy:      scaffoldRemedy("resync", relPath),
	}
	return &f
}

// scaffoldRemedy returns a ready-to-copy remedy command for a scaffold finding.
// The message is neutral about blame direction (AC16): the binary version is stated,
// but the user is told to check whether the binary or the project needs updating.
func scaffoldRemedy(action, relPath string) string {
	return fmt.Sprintf(
		"trackfw update   # %s %s: content differs from the template trackfw v%s generates; if this project was initialized with a newer binary, update the binary instead",
		action, relPath, version.Version,
	)
}

File: npm/src/integrations/scaffold_doctor.js
'use strict'

/**
 * scaffold_doctor.js — scaffold artifact coverage for `trackfw doctor`
 * (ADR-2026-08-27-doctor-cobre-artefatos-de-scaffold-por-comparacao-com-o-template-
 * com-propriedade-dada-pelo-caminho.md).
 *
 * Mirrors internal/generators/scaffold_doctor.go (Go, canonical source of truth).
 * Same design decisions apply — see that file's doc comment for the full rationale.
 *
 * Key decisions:
 *   - Property by path, not manifest (AC3): scaffold artifacts are identified by
 *     well-known namespace paths. No manifest entry is written or read.
 *   - Sibling classifier (AC15): scaffold artifacts are never in the manifest, so
 *     routing them through classifyDoctor would produce wrong remedies. The finding
 *     kinds SCAFFOLD_DIVERGENT / SCAFFOLD_MISSING are separate, with zero claim and
 *     a trackfw-update remedy.
 *   - Config-rendered templates (AC12): buildValidateScript(cfg) varies with
 *     cfg.backend/cfg.frontend — rendered from the project's own trackfw.yaml.
 *   - Eligibility for slash commands (AC14): only checked when
 *     .claude/commands/trackfw/ directory already exists.
 *   - Conditional artifacts (AC13): CI workflow only when cfg.ci declares it.
 *   - Neutral blame message (AC16): binary version stated; direction ambiguous.
 */

const fs = require('fs')
const path = require('path')
const yaml = require('yaml')

const { SIGNAL_SCRIPT, CLEANUP_SCRIPT, CREDENTIAL_GUARD_SCRIPT, GIT_BRANCH_GUARD_SCRIPT } =
  require('../generators/hooks')
const { CLAUDE_COMMANDS, buildValidateScript, buildGitHubActionsWorkflowContent, buildGitLabCIWorkflowContent } =
  require('../generators/init')

// Finding kind constants — mirrors Go's DoctorScaffoldDivergent / DoctorScaffoldMissing.
const SCAFFOLD_DIVERGENT = 'scaffold-divergent'
const SCAFFOLD_MISSING = 'scaffold-missing'

// PYTHON_VALIDATE_SCRIPT_FORM is the byte-exact content Python's `trackfw init` and
// `trackfw update` (validate-script target) write to scripts/trackfw-validate.sh.
// Accepted by the set-membership check in checkValidateScriptArtifact so that a
// project initialized by the Python runtime does not produce a false-positive
// scaffold-divergent finding in the Node.js doctor.
//
// Scope of the exception: ONLY scripts/trackfw-validate.sh uses set-membership.
// All other scaffold artifacts use single-template equality.
// See docs/cli-parity.md, "validate.sh — pertencimento a conjunto (set-membership, escopado)".
const PYTHON_VALIDATE_SCRIPT_FORM = '#!/usr/bin/env bash\nset -euo pipefail\ntrackfw validate\n'

// Paths used by the scaffold doctor (mirrors Go's exported constants).
const CLAUDE_COMMANDS_DIR_PATH = '.claude/commands/trackfw'
const GITHUB_ACTIONS_WORKFLOW_PATH = '.github/workflows/trackfw-gate.yml'
const GITLAB_CI_WORKFLOW_PATH = '.gitlab-ci-trackfw.yml'

/**
 * scaffoldRemedy returns a ready-to-copy remedy command for a scaffold finding.
 * The message is neutral about blame direction (AC16).
 */
function scaffoldRemedy(action, relPath) {
  // Lazy-require version to avoid a circular dependency at module load time.
  let ver = 'unknown'
  try {
    const pkg = require('../../package.json')
    ver = pkg.version || 'unknown'
  } catch (_) {}
  return `trackfw update   # ${action} ${relPath}: content differs from the template trackfw v${ver} generates; if this project was initialized with a newer binary, update the binary instead`
}

/**
 * loadProjectConfig reads trackfw.yaml from projectRoot and returns a cfg object
 * with the same keys that buildValidateScript / buildCIWorkflowContent expect.
 * Missing keys default to undefined/null — identical to Go's loadUpdateConfig().
 */
function loadProjectConfig(projectRoot) {
  try {
    const raw = fs.readFileSync(path.join(projectRoot, 'trackfw.yaml'), 'utf8')
    const parsed = yaml.parse(raw) || {}
    return {
      backend: parsed.backend || null,
      frontend: parsed.frontend || null,
      pkgManager: parsed.pkg_manager || null,
      ci: parsed.ci || null,
    }
  } catch (_) {
    return { backend: null, frontend: null, pkgManager: null, ci: null }
  }
}

/**
 * checkValidateScriptArtifact checks scripts/trackfw-validate.sh using set-membership:
 * accepted if the file content matches EITHER Go/Node's cfg-rendered form OR Python's
 * fixed form (PYTHON_VALIDATE_SCRIPT_FORM). A file matching NO known form is accused.
 * All other scaffold artifacts use checkScaffoldArtifact (single-template equality).
 *
 * @param {string} absPath - absolute path to the file
 * @param {string} relPath - relative path (used as destination in findings)
 * @param {object} cfg - project config from trackfw.yaml
 * @returns {object|null} finding or null
 */
function checkValidateScriptArtifact(absPath, relPath, cfg) {
  let actual
  try {
    actual = fs.readFileSync(absPath, 'utf8')
  } catch (err) {
    if (err.code === 'ENOENT') {
      return {
        finding: SCAFFOLD_MISSING,
        claim: { kind: '', item: '', target: '', surface: '', scope: '' },
        destination: relPath,
        remedy: scaffoldRemedy('restore', relPath),
      }
    }
    return {
      finding: SCAFFOLD_DIVERGENT,
      claim: { kind: '', item: '', target: '', surface: '', scope: '' },
      destination: relPath,
      remedy: scaffoldRemedy('resync', relPath),
    }
  }
  const goNodeForm = buildValidateScript(cfg)
  if (actual === goNodeForm || actual === PYTHON_VALIDATE_SCRIPT_FORM) return null
  return {
    finding: SCAFFOLD_DIVERGENT,
    claim: { kind: '', item: '', target: '', surface: '', scope: '' },
    destination: relPath,
    remedy: scaffoldRemedy('resync', relPath),
  }
}

/**
 * checkScaffoldArtifact compares the on-disk content at absPath against expected.
 * Returns a finding if the file is divergent or (when reportMissing=true) absent.
 * relPath is used as destination in the finding.
 */
function checkScaffoldArtifact(absPath, relPath, expected, reportMissing) {
  let actual
  try {
    actual = fs.readFileSync(absPath, 'utf8')
  } catch (err) {
    if (err.code === 'ENOENT') {
      if (!reportMissing) return null
      return {
        finding: SCAFFOLD_MISSING,
        claim: { kind: '', item: '', target: '', surface: '', scope: '' },
        destination: relPath,
        remedy: scaffoldRemedy('restore', relPath),
      }
    }
    // Unreadable artifact: report as divergent so the user is informed.
    return {
      finding: SCAFFOLD_DIVERGENT,
      claim: { kind: '', item: '', target: '', surface: '', scope: '' },
      destination: relPath,
      remedy: scaffoldRemedy('resync', relPath),
    }
  }
  if (actual === expected) return null
  return {
    finding: SCAFFOLD_DIVERGENT,
    claim: { kind: '', item: '', target: '', surface: '', scope: '' },
    destination: relPath,
    remedy: scaffoldRemedy('resync', relPath),
  }
}

/**
 * runScaffoldDoctor compares scaffold artifacts on disk against the templates the
 * currently installed binary would generate (given the project's own trackfw.yaml),
 * and returns findings for any artifact that is divergent or missing.
 *
 * @param {string} projectRoot - absolute path to the project root
 * @returns {Array} findings
 */
function runScaffoldDoctor(projectRoot) {
  // Eligibility: trackfw.yaml must exist.
  try {
    fs.statSync(path.join(projectRoot, 'trackfw.yaml'))
  } catch (_) {
    return []
  }

  const cfg = loadProjectConfig(projectRoot)
  const findings = []

  // --- Scripts (always in scope when trackfw.yaml is present) ---
  //
  // scripts/trackfw-validate.sh uses set-membership (Go/Node form OR Python form).
  // The four remaining scripts use single-template equality via checkScaffoldArtifact.
  const validateF = checkValidateScriptArtifact(
    path.join(projectRoot, 'scripts/trackfw-validate.sh'),
    'scripts/trackfw-validate.sh',
    cfg,
  )
  if (validateF) findings.push(validateF)

  const staticScripts = [
    { relPath: 'scripts/trackfw-attention-signal.sh', expected: SIGNAL_SCRIPT },
    { relPath: 'scripts/trackfw-attention-cleanup.sh', expected: CLEANUP_SCRIPT },
    { relPath: 'scripts/trackfw-credential-guard.sh', expected: CREDENTIAL_GUARD_SCRIPT },
    { relPath: 'scripts/trackfw-git-branch-guard.sh', expected: GIT_BRANCH_GUARD_SCRIPT },
  ]
  for (const { relPath, expected } of staticScripts) {
    const f = checkScaffoldArtifact(path.join(projectRoot, relPath), relPath, expected, true)
    if (f) findings.push(f)
  }

  // --- Slash commands (AC14: only when the directory already exists) ---
  const claudeDir = path.join(projectRoot, CLAUDE_COMMANDS_DIR_PATH)
  let claudeDirExists = false
  try {
    fs.statSync(claudeDir)
    claudeDirExists = true
  } catch (_) {}
  if (claudeDirExists) {
    for (const [filename, content] of Object.entries(CLAUDE_COMMANDS)) {
      const relPath = `${CLAUDE_COMMANDS_DIR_PATH}/${filename}`
      const f = checkScaffoldArtifact(path.join(projectRoot, relPath), relPath, content, true)
      if (f) findings.push(f)
    }
  }

  // --- CI workflow (AC13: conditional on ci: in trackfw.yaml) ---
  if (cfg.ci === 'github-actions') {
    const relPath = GITHUB_ACTIONS_WORKFLOW_PATH
    const f = checkScaffoldArtifact(
      path.join(projectRoot, relPath),
      relPath,
      buildGitHubActionsWorkflowContent(cfg),
      true,
    )
    if (f) findings.push(f)
  } else if (cfg.ci === 'gitlab-ci') {
    const relPath = GITLAB_CI_WORKFLOW_PATH
    const f = checkScaffoldArtifact(
      path.join(projectRoot, relPath),
      relPath,
      buildGitLabCIWorkflowContent(cfg),
      true,
    )
    if (f) findings.push(f)
  }

  // Deterministic output (AC7): sort by destination.
  findings.sort((a, b) => (a.destination < b.destination ? -1 : a.destination > b.destination ? 1 : 0))

  return findings
}

module.exports = {
  runScaffoldDoctor,
  checkValidateScriptArtifact,
  SCAFFOLD_DIVERGENT,
  SCAFFOLD_MISSING,
  PYTHON_VALIDATE_SCRIPT_FORM,
}
